rfcs/0004-categorized-attributes.md

```diff
@@ -0,0 +1,217 @@
+
+An implementation MUST allow efficient query of all _categorized attributes_ when given a _category_. It MUST allow for efficient transformation of all _categorized attributes_ of a given category, such as adding another set of matching attributes to it on a per-value basis.
+
+For instance, such an operation MUST be performed efficiently:
+
+```
+capacity
+	shield 100
+	hull 20
+
+-
+
+damage
+	shield 10
+	hull 10
+
+=
+
+capacity
+	shield 90
+	hull 10
+```
+
+An implementation SHOULD store _categorized attributes_ grouped by _category_. It SHOULD NOT store the _category_ of the individual attribute values.
+
+An implementation MAY use a similar structure for storing attributes:
+
+```
+class Entry {
+	Effect e;
+	Modifier m;
+}
+
+std::map<Entry, double> attributes[CATEGORY_COUNT];
+```
+```
+class Entry {
+	Effect e;
+	Modifier m;
+
+	double value;
+}
+
+std::map<Category, std::set<Entry>> attributes;
+```
+
+An implementation MUST provide const iterator functionality over attributes. It MAY provide non-const iterator functionality.
+
+An implementation MUST provide efficient functionality matching that of Outfit::CanAdd for _categorized attributes_.
+
+# Drawbacks
+
+## Maintenance cost:
+
+We have to keep a list of all _old-form_ attribute names that the engine refers to using the new format. Though this list need not be updated or maintained, it represents an additional cost.
+
+This change will also heavily conflict with and other pull requests that create new _categorized attributes_, or change engine code using such attributes. They will all have to be moved to the _new-form_ syntax, creating a burden on the PR authors.
+
+## Performance
+
+Though the performance of the final form of this RFC should be better than our current performance, it would perform significantly worse while only the data format is implemented. Early tests suggest a doubling of the individual lookup cost of categorized attributes using the currently implemented method in the worst case.
+
+# Alternatives
+
+We could programmatically generate the _old-form_ attribute names. This would preserve a larger portion of the existing code base and would require no changes from plugin authors, however [past efforts show it to be prohibitively complex](https://github.com/tibetiroka/endless-sky/blob/99108ba344a9de01bcab5eba7f949e39b9bb3758/source/Attribute.cpp#L85-L165).
+
+# Unresolved Questions
+
+Is there a better way of managing compatibility than storing the _old-form_ to _new-form_ mappings?
+
+Is it worth using a Dictionary-style flap map over std::map for storing attributes? Is it really that much faster even for non-_categorized attributes_?
+
+Should the constraints of attribute values be enforced by the individual attribute entries, or the data store? Is the performance cost of enforcing them on every operation worth it? In what cases should we validate the input?
```

Comment:
Validation can happen at load-time. So it should be okay to validate whatever we want to validate at load-time.